Lib/tarfile.py

```diff
@@ -707,9 +714,127 @@ def __init__(self, tarfile, tarinfo):
+        self._path = path
+        super().__init__(f'{tarinfo.name!r} would be extracted to {path!r}, '
+                         + 'which is outside the destination')
+
+class SpecialFileError(FilterError):
+    def __init__(self, tarinfo):
+        self.tarinfo = tarinfo
+        super().__init__(f'{tarinfo.name!r} is a special file')
+
+class AbsoluteLinkError(FilterError):
+    def __init__(self, tarinfo):
+        self.tarinfo = tarinfo
+        super().__init__(f'{tarinfo.name!r} is a symlink to an absolute path')
+
+class LinkOutsideDestinationError(FilterError):
+    def __init__(self, tarinfo, path):
+        self.tarinfo = tarinfo
+        self._path = path
+        super().__init__(f'{tarinfo.name!r} would link to {path!r}, '
+                         + 'which is outside the destination')
+
+def _get_filtered_attrs(member, dest_path, for_data=True):
+    new_attrs = {}
+    name = member.name
+    dest_path = os.path.realpath(dest_path)
+    # Strip leading / (tar's directory separator) from filenames.
+    # Include os.sep (target OS directory separator) as well.
+    if name.startswith(('/', os.sep)):
+        name = new_attrs['name'] = member.path.lstrip('/' + os.sep)
+    if os.path.isabs(name):
+        # Path is absolute even after stripping.
+        # For example, 'C:/foo' on Windows.
+        raise AbsolutePathError(member)
+    # Ensure we stay in the destination
+    target_path = os.path.realpath(os.path.join(dest_path, name))
+    if os.path.commonpath([target_path, dest_path]) != dest_path:
+        raise OutsideDestinationError(member, target_path)
+    # Limit permissions (no high bits, and go-w)
+    mode = member.mode
+    if mode is not None:
+        # Strip high bits & group/other write bits
+        mode = mode & 0o755
+        if for_data:
+            # For data, handle permissions & file types
+            if member.isreg() or member.islnk():
+                if not mode & 0o100:
+                    # Clear executable bits if not executable by user
+                    mode &= ~0o111
+                # Ensure owner can read & write
+                mode |= 0o600
+            elif member.isdir() or member.issym():
+                # Ignore mode for directories & symlinks
+                mode = None
+            else:
+                # Reject special files
+                raise SpecialFileError(member)
+        if mode != member.mode:
+            new_attrs['mode'] = mode
+    if for_data:
+        # Ignore ownership for 'data'
+        if member.uid is not None:
+            new_attrs['uid'] = None
+        if member.gid is not None:
+            new_attrs['gid'] = None
+        if member.uname is not None:
+            new_attrs['uname'] = None
+        if member.gname is not None:
+            new_attrs['gname'] = None
+        # Check link destination for 'data'
+        if member.islnk() or member.issym():
+            if os.path.isabs(member.linkname):
+                raise AbsoluteLinkError(member)
```

Comment:
On Windows, `os.path.isabs()` incorrectly classifies rooted paths (e.g. "\\spam") as absolute paths. They're actually relative paths since they depend on the drive or share of either the current working directory or the opened path of a symlink.

In cases such as this, however, rooted paths need to be excluded even though they're technically relative paths on Windows. If we fixed `ntpath.isabs()`, which would fix bugs caused by the current behavior, it would break this use case. An alternative for this case would be to use `drive, root, rest = os.path.splitroot(member.linkname)`, and raise `AbsoluteLinkError` if either `drive` or `root` is non-empty.

Reply:
The `\spam` case would be caught in ihe next check, where `tarfile` ensures that it's extracting into the destination directory. That check is necessary for relative paths (like `..\spam`), anyway.
The `AbsolutePathError` vs. `OutsideDestinationError` is more of a hint about what's wrong with the archive, and IMO it's OK to follow the current behavior of `isabs`.(An attacker with knowledge of the unpacking environment *could* craft an archive with 

There is no `os.path.splitroot`, should the proper solution be `splitdrive` followed by `isabs`?

Could you adjust the `isabs` docs of to point out the issue and suggest the alternative?

Reply:
The `\spam` case would be caught in ihe next check, where `tarfile` ensures that it's extracting into the destination directory. That check is necessary for relative paths (like `..\spam`), anyway.
The `AbsolutePathError` vs. `OutsideDestinationError` is more of a hint about what's wrong with the archive, and IMO it's OK to follow the current behavior of `isabs`.
You can craft an archive that can only be extracted into a specific directory, with paths like like `..\dest\foo.txt` (or `\some\deep\path\foo.txt` after `isabs` is fixed), but IMO that's not a *security* issue.

There is no `os.path.splitroot`, should the proper solution be `splitdrive` followed by `isabs`?

Could you adjust the `isabs` docs of to point out the issue and suggest the alternative?

Reply:
That's exactly the kind of code I'd like to avoid maintaining in `tarfile`.
If there was a `os.path.is_absolute`, and `os.path.isabs` docs said to use that instead, I'll be happy to do it. But platform-specific path manipulation is too fragile to put in a security-sensitive module with no active expert.


The difference between `AbsoluteLinkError` and `LinkOutsideDestinationError` is that the first is more likely to be caused by misconfigured archival software, while the latter is more likely to be a malicious attack/prank or an issue with a single stray path. IMO it's good to use different exceptions, but again, the edge cases don't matter that much.